File: consoles/cafe.py
import shutil
from pathlib import Path

from console import Console


class Cafe(Console):
    Elf2Rpl = 'elf2rpl "{elf}" "{output}.rpx"'
    BinTool = 'wuhbtool "{rpx}" "{output}.wuhb" --content="{romfs}" --name="{short_name}" --short-name="{short_name}" --author="{author}" --icon="{icon}"'

    def __init__(self, metadata: dict) -> None:
        super().__init__(metadata)

    def build(self, build_dir: Path) -> str:
        shutil.copy(self.game_zip, self.path_to("content"))

        args = {
            "elf": self.binary_path(),
            "output": build_dir / self.title,
        }

        error = self.run_command(Cafe.Elf2Rpl, args)

        if error != "":
            return error

        args = {
            "rpx": build_dir / f"{self.title}.rpx",
            "output": build_dir / self.title,
            "romfs": self.path_to("content"),
            "name": self.title,
            "short_name": self.title,
            "author": self.author,
            "icon": self.icon_file(),
        }

        error = self.run_command(Cafe.BinTool, args)

        if error != "":
            return error

        Path(self.path_to("content/game.zip")).unlink()

        return str()

    def binary_extension(self) -> str:
        return "wuhb"

    def icon_extension(self) -> str:
        return "png"

File: server/src/types/console/cafe.rs
use std::path::Path;

use crate::enums::api_error::ApiError;
use crate::enums::console::Console;
use crate::modules::filesystem;
use crate::traits::binary::Binary;
use crate::types::bundler_query::Metadata;

pub struct Cafe {}

const RPL_TOOL: &str = "elf2rpl";
const BINARY_TOOL: &str = "wuhbtool";

impl Binary for Cafe {
    async fn compile(
        basepath: &Path,
        metadata: &Metadata,
        icon: &String,
    ) -> Result<String, ApiError> {
        let output_path = filesystem::normalize_path(basepath);
        let binary_path = Console::Hac.get_binary();

        Self::run_command(RPL_TOOL, &[&binary_path, &format!("{}.rpx", output_path)]).await?;

        Self::run_command(
            BINARY_TOOL,
            &[
                &format!("{}.rpx", output_path),
                &format!("{}.wuhb", output_path),
                &format!("--content={}", Console::Cafe.get_romfs()),
                &format!("--name={}", metadata.title),
                &format!("--short-name={}", metadata.title),
                &format!("--author={}", metadata.author),
                &format!("--icon={}", icon),
            ],
        )
        .await?;

        let base64 = filesystem::read_file_as_base64(&format!("{}.wuhb", output_path)).await?;

        Ok(base64)
    }
}
